### PR TITLE
GEODE-3637: Reimplement client queue initialization. Adding shutdown …

### PR DESCRIPTION
**Pull request by upthewaterspout**
…logic

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

@PivotalSarge 
@WireBaron

### CODE REVIEW

File: geode-core/src/test/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImplClientQueueDUnitTest.java [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+  public void tearDown() throws RemoteException {
+    blackboard.initCount();
+    host.getAllVMs().forEach((vm) -> vm.invoke(() -> {
+      InitialImageOperation.slowImageProcessing = 0;
+      System.getProperties().remove("BridgeServer.HANDSHAKE_POOL_SIZE");
+    }));
+  }
+
+  @Test
+  public void testClientSubscriptionQueueBlockingConnectionInitialization() throws Exception {
```
Comment (jhuynh1) [inline geode-core/src/test/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImplClientQueueDUnitTest.java]: Is this test making sure our queues block during initialization?

Just had a hard time figuring out what the test was actually doing

File: geode-core/src/test/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImplClientQueueDUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+          cache.readyForEvents();
+          Awaitility.await().atMost(10, TimeUnit.SECONDS).pollInterval(1, TimeUnit.SECONDS)
+              .until(() -> eventCount.get() == numberOfEntries);
+          cache.close();
+          return eventCount.get() == numberOfEntries;
+        });
+
+    vm3.invokeAsync("Start Client2 to add entries to region", () -> {
+      while (true) {
+        Thread.sleep(100);
```
Comment (jhuynh1) [inline geode-core/src/test/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImplClientQueueDUnitTest.java]: Possible timing issue?

**Comment by PivotalSarge** [comment]
If memory serves, the original intention was to prevent the handshaking from blocking the acceptor thread by moving it to its own thread. Clients should still need to follow the same initialization sequence as before.

Sarge

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java [hunk 3]
```diff
@@ -1202,7 +1208,8 @@ public SelectableChannel getSelectableChannel() {
   }
 
   public void registerWithSelector2(Selector s) throws IOException {
-    /* this.sKey = */getSelectableChannel().register(s, SelectionKey.OP_READ, this);
+    /* this.sKey = */
```
Comment (PivotalSarge) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java]: Die, out-commented code! Die!

**Comment by upthewaterspout** [comment]
@jhuynh1 - I think a client could connect and do puts before it's queue was ready with or without these changes. Puts go over a different connection.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImpl.java [hunk 4]
```diff
@@ -1498,6 +1493,17 @@ protected void handleNewClientConnection(final Socket socket,
     }
   }
 
+  private boolean initializeClientPools(Socket socket, CommunicationMode communicationMode) {
```
Comment (WireBaron) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImpl.java]: This method name doesn't seem to quite capture the functionality, maybe something like handOffClientQueueInitialization?

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImpl.java [hunk 5]
```diff
@@ -534,103 +540,126 @@ public AcceptorImpl(int port, String bindHostName, boolean notifyBySubscription,
...
-      };
-      try {
-        final BlockingQueue bq = new SynchronousQueue();
-        final RejectedExecutionHandler reh = new RejectedExecutionHandler() {
-          public void rejectedExecution(Runnable r, ThreadPoolExecutor pool) {
+        };
+        return new Thread(clientQueueThreadGroup, runnable, threadName);
+      }
+    };
+    return new PooledExecutorWithDMStats(new SynchronousQueue(), 16, getStats().getCnxPoolHelper(),
```
Comment (WireBaron) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImpl.java]: Can we move some of these magic numbers to class constants.

File: geode-core/src/test/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImplClientQueueDUnitTest.java (hunk 1 above)
Comment (upthewaterspout) [reply geode-core/src/test/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImplClientQueueDUnitTest.java]: I think it's trying to make sure the queues don't block. I'll clean up the test and add some more comments.

File: geode-core/src/test/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImplClientQueueDUnitTest.java (hunk 2 above)
Comment (upthewaterspout) [reply geode-core/src/test/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImplClientQueueDUnitTest.java]: Hmm, maybe, but looking at the code this whole blackboard and while loop is not guaranteed to do anything useful. I'll clean this up.

**Pull request merged**